### DISCUSSION TITLE
Why was support for Ubuntu 14.04 dropped?

### DISCUSSION THREAD

**Discussion by bezelebub** [gravitational/teleport#34200] [Q&A] [answered]
I'm in the process of implementing teleport in my company.

Unfortunately, we do have a bunch of old systems running Ubuntu 14.04 32bit that we need to manage, but cant update for a list of reasons...

Therefore I have been trying to compile a version from source that is statically linked, and i have been somewhat successful, the binary runs and from what i could tell libc.so.6 was to old on the ubuntu 14 systems and that is why the precompiled binary didn't launch.
 
 But im running into weird issues, so I'm asking if there is something with Ubuntu 14.04 that prevents the teleport binary from working as expected and this is why support was dropped, or should it technically be possible to run in on those systems?
 
 
 To give some context of the "weird issues" here is a log snippet of an attempt to connect to the node:
 ```
023-11-03T13:05:49+01:00 DEBU             Will create new session for SSH connection 192.168.69.106:56654. srv/ctx.go:660
2023-11-03T13:05:49+01:00 DEBU [NODE]      Handling request pty-req, want reply true. id:1 local:192.168.65.172:37063 login:acg remote:192.168.69.106:56654 teleportUser:anal regular/sshserver.go:1607
2023-11-03T13:05:49+01:00 DEBU [NODE]      Requested terminal "xterm-256color" of size {125 67} id:1 local:192.168.65.172:37063 login:acg remote:192.168.69.106:56654 teleportUser:anal srv/termhandlers.go:78
2023-11-03T13:05:49+01:00 DEBU             Set permissions on /dev/pts/2 to 1000:5 with mode -rw-------. srv/term.go:484
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Unable to update window size, no session found in context. srv/sess.go:553
2023-11-03T13:05:49+01:00 DEBU             Will create new session for SSH connection 192.168.69.106:56654. srv/ctx.go:660
2023-11-03T13:05:49+01:00 DEBU [NODE]      Handling request shell, want reply true. id:1 local:192.168.65.172:37063 login:acg remote:192.168.69.106:56654 teleportUser:anal regular/sshserver.go:1607
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Tracking participant: f536e766-c153-433e-804f-82b2c7e05786 session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:2038
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Attempting to create session tracker session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:2076
2023-11-03T13:05:49+01:00 INFO [NODE]      Creating (interactive) session 967fb132-86fc-48f0-91c4-6fe91b59f68a. id:1 local:192.168.65.172:37063 login:acg remote:192.168.69.106:56654 teleportUser:anal srv/sess.go:339
2023-11-03T13:05:49+01:00 INFO [SESSION:N] New party ServerContext(192.168.69.106:56654->192.168.65.172:37063, user=acg, id=1) party(id=f536e766-c153-433e-804f-82b2c7e05786) joined the session with participant mode: peer. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1857
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Launching session session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1183
2023-11-03T13:05:49+01:00 INFO [AUDIT]     session.start addr.remote:192.168.69.106:56654 cluster_name:acg code:T2000I ei:0 event:session.start initial_command:[] login:acg namespace:default proto:ssh server_hostname:14test server_id:595eebc4-3b98-45b3-a660-46bc336526cc customer:acg environment:showroom family:Q session_recording:proxy sid:967fb132-86fc-48f0-91c4-6fe91b59f68a size:125:67 time:2023-11-03T12:05:49.336Z uid:2eeed862-76f7-41b4-b3e8-ab0d65735faa user:anal events/emitter.go:271
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Waiting for continue signal session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1308
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Got continue signal session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1313
2023-11-03T13:05:49+01:00 DEBU             Will join session 967fb132-86fc-48f0-91c4-6fe91b59f68a for SSH connection 192.168.69.106:56654. srv/ctx.go:658
2023-11-03T13:05:49+01:00 DEBU [NODE]      Exec request ("/proc/self/exe") complete: 2 id:1 local:192.168.65.172:37063 login:acg remote:192.168.69.106:56654 teleportUser:anal regular/sshserver.go:1587
2023-11-03T13:05:49+01:00 INFO [SESSION:N] Closing party f536e766-c153-433e-804f-82b2c7e05786 srv/sess.go:2026
2023-11-03T13:05:49+01:00 INFO [SESSION:N] Removing party ServerContext(192.168.69.106:56654->192.168.65.172:37063, user=acg, id=1) party(id=f536e766-c153-433e-804f-82b2c7e05786) from session session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1526
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] No longer tracking participant: f536e766-c153-433e-804f-82b2c7e05786 session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1534
2023-11-03T13:05:49+01:00 DEBU [NODE]      Releasing associated resources - context has been closed. id:1 local:192.168.65.172:37063 login:acg remote:192.168.69.106:56654 teleportUser:anal srv/monitor.go:397
2023-11-03T13:05:49+01:00 INFO [AUDIT]     session.leave cluster_name:acg code:T2003I ei:0 event:session.leave login:acg namespace:default server_hostname:14test server_id:595eebc4-3b98-45b3-a660-46bc336526cc customer:acg environment:showroom family:Q sid:967fb132-86fc-48f0-91c4-6fe91b59f68a time:2023-11-03T12:05:49.462Z uid:bba9cec0-f43a-4509-92d6-73782179b93b user:anal events/emitter.go:271
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Session has no active party members. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1598
2023-11-03T13:05:49+01:00 DEBU [SESSION:N] Copying from Party f536e766-c153-433e-804f-82b2c7e05786 to session writer completed with error <nil>. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1866
2023-11-03T13:05:49+01:00 DEBU [SSH:NODE]  Closed connection 192.168.69.106:56654. sshutils/server.go:515
2023-11-03T13:05:54+01:00 ERRO [SESSION:N] Timed out waiting for PTY copy to finish, session data  may be missing. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1335
2023-11-03T13:05:54+01:00 INFO [SESSION:N] Stopping session session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:859
2023-11-03T13:05:54+01:00 DEBU [TERM:LOCA] Closing TTY srv/term.go:339
2023-11-03T13:05:54+01:00 DEBU [SESSION:N] Copying from PTY to writer completed with error read /dev/ptmx: input/output error. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1228
2023-11-03T13:05:54+01:00 INFO [AUDIT]     session.end addr.remote:192.168.69.106:56654 cluster_name:acg code:T2004I ei:0 enhanced_recording:false event:session.end interactive:true login:acg namespace:default participants:[anal] proto:ssh server_hostname:14test server_id:595eebc4-3b98-45b3-a660-46bc336526cc customer:acg environment:showroom family:Q session_recording:proxy session_start:2023-11-03T12:05:49.331947944Z session_stop:2023-11-03T12:05:54.460847241Z sid:967fb132-86fc-48f0-91c4-6fe91b59f68a time:2023-11-03T12:05:54.461Z uid:62ba7ddf-e1cf-47c2-a837-abc64b5650e6 user:anal events/emitter.go:271
2023-11-03T13:05:54+01:00 DEBU [SESSION:N] Copying from reader to PTY completed with error <nil>. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1236
2023-11-03T13:05:54+01:00 DEBU [TERM:LOCA] Closed TTY srv/term.go:357
2023-11-03T13:05:54+01:00 DEBU [TERM:LOCA] Terminal child process has been stopped srv/term.go:289
2023-11-03T13:05:54+01:00 INFO [SESSION:N] Closing session session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:908
2023-11-03T13:05:54+01:00 DEBU [TERM:LOCA] Closed PTY srv/term.go:377
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Runtime config: restrict_connection_count: false connection_count: 1 localCluster: targetCluster:acg reversetunnel/agentpool.go:335
2023-11-03T13:05:55+01:00 DEBU             Attempting GET teleport.q-web.se:443/webapi/find webclient/webclient.go:129
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Starting agent {teleport.q-web.se:443 tcp } leaseID:18 target:teleport.q-web.se:443 localCluster: targetCluster:acg reversetunnel/agent.go:297
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Changing state initial -> connecting. leaseID:18 target:teleport.q-web.se:443 localCluster: targetCluster:acg reversetunnel/agent.go:285
2023-11-03T13:05:55+01:00 DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:195
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Aborting SSH handshake because the proxy "eba99901-61ad-4f62-9b88-80e0b9fcba17.acg" is already claimed by some other agent. localCluster: targetCluster:acg reversetunnel/agent_dialer.go:76
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Changing state connecting -> closed. leaseID:18 target:teleport.q-web.se:443 localCluster: targetCluster:acg reversetunnel/agent.go:285
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Failed to connect agent: ssh: handshake failed: proxy already claimed. localCluster: targetCluster:acg reversetunnel/agentpool.go:274
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Active agent count: 1 localCluster: targetCluster:acg reversetunnel/agentpool.go:418
2023-11-03T13:05:55+01:00 DEBU [PROXY:AGE] Active agent count: 1 localCluster: targetCluster:acg reversetunnel/agentpool.go:418
```

i followed the suggestions here to compile statically:
[https://github.com/gravitational/teleport/issues/4896](url)

I also tested the same binary on a Ubuntu 22 system and it works as expected

**Comment by zmb3** [comment]
I don't know of any reason why Teleport would not work on 14.04 if you compile from source with a compatible version of libc. 

I don't see any errors in the logs that you've shared. Looks like standard debug logs to me.

**Reply by bezelebub** [reply]
Hi, Thanks for you awnser,

Perhaps the logdump is a bit unclear, i was trying to provide some context.
The actual error in the logs is: 
`2023-11-03T13:05:54+01:00 ERRO [SESSION:N] Timed out waiting for PTY copy to finish, session data  may be missing. session_id:967fb132-86fc-48f0-91c4-6fe91b59f68a srv/sess.go:1335`

A bit hidden without colour highligthing. 
To be honest im out of my dephth here, but it seems that there is a problem creating a pty session, and that results in the node trying to connect again but the proxy is the already in use.

I have seen similar behavior in the past with teleport but for other reasons.

As compiling isent really my normal expertiese im looking for some guidence on what might be wrong or best practises.
The binary im currently working with was cross compiled on a ubuntu 22 64 bit system for 32bit and statically linked to include the newer libc. Is there a better way?

I first tried to compile on a Ubuntu 14 system, but ran into to many issues sence all teleports dependencys also dont support 32 bit.

So to sum up the 2 questions i have:
What is the best approch to compiling for Ubuntu 14 32 bit, (or any older not offically supported distro)?

And given the issues with the current binary, do you have any suggestions on what to start looking at given the error and behavior?

Thanks for you time and apologies for any wrong assumptions i might have made.
Id also be happy to provide more details if needed

**Comment by bezelebub** [comment, marked as answer]
Update,

I got it to work, with some help from chatgpt i was able to compile teleport for an older version of libc on a ubuntu 22 system using a tool called crosstool-ng to setup a enviroment for compining a dynamic binary.

This compiled version still produces the same looking log, but connecting to the node actually works. 
That it did not work with the statically linked binary.
